api/v1/client.go

```diff
@@ -653,7 +657,56 @@ func (c *storageClient) RuntimeTransactions(ctx context.Context, r *http.Request
 		return nil, common.ErrBadRequest
 	}
 
-	return c.storage.RuntimeTransactions(ctx, &q, &p)
+	sts, err := c.storage.RuntimeTransactions(ctx, &q, &p)
+	if err != nil {
+		return nil, err
+	}
+
+	var ts RuntimeTransactionList
+	for _, st := range sts.Transactions {
+		var utx types.UnverifiedTransaction
+		if err = cbor.Unmarshal(st.Raw, &utx); err != nil {
+			return nil, fmt.Errorf("round %d tx %d utx unmarshal: %w", st.Round, st.Index, err)
+		}
+		tx, err := uncategorized.OpenUtxNoVerify(&utx)
+		if err != nil {
+			return nil, fmt.Errorf("round %d tx %d utx open no verify: %w", st.Round, st.Index, err)
+		}
+		sender0, err := uncategorized.StringifyAddressSpec(&tx.AuthInfo.SignerInfo[0].AddressSpec)
+		if err != nil {
+			return nil, fmt.Errorf("round %d tx %d signer 0: %w", st.Round, st.Index, err)
+		}
+		var cr types.CallResult
+		if err = cbor.Unmarshal(st.ResultRaw, &cr); err != nil {
+			return nil, fmt.Errorf("round %d tx %d result unmarshal: %w", st.Round, st.Index, err)
+		}
```

Comment:
I wanted to have it done in the client originally, back in the nexus project. but later the scanner and server for the backend were done before the frontend work started, so I added some backend-side parsing so that we can move faster when the frontend development starts.

it does seem more oasis-indexer-like to parse fully while scanning though. and it would be nice to restructure all this to be that way.

other than these historical reasons, I do have arguments for having designed it this way, but we don't have to prioritize them above following oasis-indexer style.

zero: you use the phrase "[parse things in the scanner]; that's why we have analyzers" to say that, at least in the scope of the indexer project "analyzer" should mean that it does more extensive processing. let's go with that for now. in order to discuss this without raising a dispute about terminology, I'll be using "scanner" to refer to the component that processes blocks as they are created, without any qualification about how extensive that processing is. the scanner in oasis-indexer's consensus side is an analyzer. the scanner in nexus's emerald side was not an analyzer.

1. the more we do on the server side and less we do on the scanner side, the more development we can do without needing a rescan
2. if the output of the scanning has a schema that's tightly coupled with the oasis-core/oasis-sdk types, e.g. a column for each field, it's especially tempting to change the schema when oasis-core/oasis-sdk changes, which will need a big migration and/or rescan
3. the genesis migration tool in oasis-core demonstrates that it is possible to write a program that works with multiple versions of structures
4. if all analysis is mixed into the scanner, it'll take some effort to track backwards-compatible changes and changes that need a rescan

so it was more appealing to design a scanner that would do a specific small set of work. the plan for nexus was a lot smaller scope after all, compared to oasis-indexer being "this is the only scanner. oasis-indexer must support all use cases present and future." 

> It lets us perform ad-hoc queries/analysis on the DB, and it leaves the web server more lightweight (= harder to DoS).

the remaining analysis to be done in the server is designed to be lightweight. we don't even verify a signature anymore. ugh do we even have a sense of how much the cbor unmarshalling etc costs compared to the rest of the overhead in responding to a request?

Reply:
Fair enough that doing more processing on the fly made more sense for the previous, slightly narrower project scope.

I still think preprocessing brings usability and robustness benefits because indexer's entire functionality can be checked from just the DB. Debugging somebody's missing tx with a one-off query, performing a db join on the transactions table that was not anticipated in advance (and not having to do tx postprocessing/analysis), being able to index whichever field we want ... Also, it will force us to run the deep analysis on _all_ the txs, so if there's a problem (malformed input, violated assumption, etc), we'll learn about it right away, not "at runtime" when somebody queries the tx.

Ack that we do pay the price of more frequent re-scans. I don't see re-scans as a big problem; I would argue they might even be beneficial. Because the indexer (and Explorer) are such key parts of the overall infrastructure, we should become well-versed at re-scanning (with either a bugfix or a new feature) at a whim. If it's an arcane process that we do once a quarter, emergency patches become riskier. 

As an alternative to full re-scans, I have also been thinking about incremental re-scans (i.e. having scanners/analyzers in multiple stages, each stage building from the previous), but I'd prefer not to go there until demonstrably needed because it's a more complex setup.

All that said, with our 0 current usage, most of our arguments are somewhat hypothetical/anticipatory, so I won't fight this more. If I haven't convinced you enough and you want to keep the on-the-ly processing, please let's just implement it so it's easy to migrate to an analyzer later. Meaning I'd like us to have a func from `storage.RuntimeTransaction` to `api.RuntimeTransaction`. If you are ok to call it in the analyzer instead of the client, even better.

Reply:
although a few weeks ago I was in favor of restructuring the processing to take place at scan time, I've changed my mind.

I'm reviewing some of our future plans, and we'll be taking on several sources of data other than data from the chain. we'll need to combine that with the data we get during this chain scanning to serve some APIs, for example, to serve parsed EVM events.

we don't have much time to deliver all these upcoming features, and I think we'll be able to move faster if we keep the database simple, with as few things stored as needed. storing more is akin to long-term caching api results, and I'm afraid of us having a hard time getting the invalidation right when we have more data sources.

so I'll be taking you up on the offer not to move all the processing to the scan step at this time.

---

> indexer's entire functionality can be checked from just the DB

would be nice. I like static site generators for a comparable reason. although I've been using a static site generators that has to rebuild everything for each change, ugh

> Debugging somebody's missing tx with a one-off query

haven't tried. I think I don't see this being as valuable as you do

> performing a db join on the transactions table that was not anticipated in advance, being able to index whichever field we want

would be nice. but we will be able to rate the work to add those columns as low complexity

> it will force us to run the deep analysis on all the txs

would be nice. although that's not intrinsic to doing the processing during scanning. we may choose to test that in either design. and I feel like doing the processing on the api side will make the production indexer more likely to fail gracefully

Reply:
agreed on extracting this from the http handler though, as we'll definitely need it for the single-transaction-detail api too

Reply:
:+1: to everything.
Especially agreed that processing on the fly will help with development velocity, and that's becoming increasingly important.

Also, I've meanwhile found another argument against being too gung-ho on frequent reindexes: info from archive nodes is awkward to handle.

Reply:
extracted api/v1/logic.go renderRuntimeTransaction